app/gui/view/component-browser/breadcrumbs/src/lib.rs

```diff
@@ -0,0 +1,478 @@
+        let text = main.create_sublayer();
+        main.set_mask(&mask);
+        Layers { main, text, mask }
+    }
+}
+
+
+
+// =============
+// === Model ===
+// =============
+
+/// A breadcrumbs model.
+#[derive(Debug, Clone, CloneRef)]
+pub struct Model {
+    display_object: display::object::Instance,
+    grid:           GridView,
+    entries:        Entries,
+    network:        frp::Network,
+    mask:           mask::View,
+    show_ellipsis:  Rc<Cell<bool>>,
+}
+
+impl Model {
+    /// Constructor.
+    pub fn new(app: &Application) -> Self {
+        let display_object = display::object::Instance::new(&app.logger);
+        let mask = mask::View::new(&app.logger);
+        display_object.add_child(&mask);
+        let grid = GridView::new(app);
+        let params = entry::Params::default();
+        grid.set_entries_params(params);
+        grid.reset_entries(1, 0);
+        display_object.add_child(&grid);
+        let entries: Entries = default();
+        let show_ellipsis = Rc::new(Cell::new(false));
+        frp::new_network! { network
+            requested_entry <- grid.model_for_entry_needed.map2(&grid.grid_size,
+                f!([entries, show_ellipsis]((row, col), grid_size) {
+                    let (_, cols) = grid_size;
+                    (*row, *col, Self::entry_model(&entries, *col, show_ellipsis.get(), *cols))
+                })
+            );
+            grid.model_for_entry <+ requested_entry;
+        }
+        Self { display_object, grid, entries, network, mask, show_ellipsis }
+    }
+
+    fn set_layers(&self, layers: Layers) {
+        layers.mask.add_exclusive(&self.mask);
+        layers.main.add_exclusive(&self.display_object);
+        layers.main.add_exclusive(&self.grid);
+        self.grid.set_text_layer(Some(layers.text.downgrade()));
+        self.network.store(&layers);
+    }
+
+    /// Update the displayed height of the breadcrumbs entries.
+    fn update_entries_height(&self, height: f32) {
+        // The width of the entries is unimportant because each entry will adjust the columns' sizes
+        // according to its content.
+        let width = 0.0;
+        self.grid.set_entries_size(Vector2(width, height));
+    }
+
+    /// Update the position and the viewport of the underlying Grid View. If the content does not
+    /// fit into [`size`], it is cropped by the rectangular [`mask`] and shifted left. So that
+    /// the user always sees the right (most important) part of the breadcrumbs.
+    fn update_layout(&self, content_size: Vector2, size: Vector2) {
+        self.mask.size.set(size);
+        let grid_view_center = Vector2(size.x / 2.0, -size.y / 2.0);
+        self.mask.set_position_xy(grid_view_center);
+        // Additional padding is added to avoid rare glitches when the last entry is
+        // cropped because it is placed right on the border of the viewport. Even 1px seems
+        // enough, but we add a bit more to be sure.
+        let padding = 10.0;
```

Comment:
what is this padding for? The docs dont explain it - it is padding between what and what?

Reply:
Adjusted comment. It is padding between the content and the right border of the viewport.